WIP: use deno to compile to single binary

6543 opened the pull request:
based on https://github.com/woodpecker-ci/plugin-ready-release-go/pull/609

```
 deno compile --allow-read --allow-write --allow-env --allow-net --allow-run src/run.ts
```


https://deno.com/blog/v1.46

6543 commented:
![image](https://github.com/user-attachments/assets/29e7d6c8-d169-463d-bf2b-eff3ae516020)

well now i just need to make sure the git impl. is working currectly and cant be broken by user input ...

... but for my 1/2h experiment, it should be doable

6543 commented on src/utils/git.ts:
```diff
@@ -0,0 +1,139 @@
+  push(args: string[]): Promise<void>;
+}
+
+export function simpleGit(): SimpleGit {
```
mostly untested & we should add tests for it anyway ...

pat-s commented:
Will this also reduce the image size? Could we even use a scratch image?

6543 commented:
not scratch _jet_ as we need the git binary working

but yes that's a way to move towards it :)

6543 commented:
the image size would probably increase by ~50Mb but the image would be silli and dont contain any unessesary build-time deps and other unrelated stuff

6543 commented:
tldr: if we have a static linked git binary + the deno binary we should be able to use FROM scratch

pat-s commented:
```
git clone git@github.com:git/git.git
mkdir git-static
cd git
make configure
./configure prefix=/path/to/git-static/ CFLAGS="${CFLAGS} -static"
make
make install
```

claude committed:
feat: compile plugin to single binary with deno compile

Deno 2's Node compatibility runs the existing tsx/Node sources
unchanged, so a hand-rolled git wrapper (as attempted in #186 via the
removed Deno.run API) is unnecessary; simple-git and process.env keep
working as-is. deno.json carries Deno's own compilerOptions since it
misreads the Node-oriented tsconfig, and the compile task skips
type-checking because tsc already covers it and @types/node is absent
from production installs. Runtime loading of release-config.ts from
disk keeps working inside the compiled binary. Installing with --prod
before compiling shrinks the embedded node_modules from ~100MB to
~21MB.
Refs #186

6543 committed:
use explicit .ts extensions for relative

claude committed:
feat: manage dependencies and tooling with deno instead of pnpm

Deno resolves npm dependencies straight from package.json and locks
them in deno.lock, so pnpm, corepack and node are no longer needed for
development. tsx is replaced by deno run, tsc by deno check (which
ships its own node types, dropping typescript and @types/node), and
vitest and prettier keep running unchanged through deno run npm:. The
tsconfig moves into deno.json compilerOptions. CI switches from the
node image with corepack to the deno image running the same tasks.

claude committed:
feat: build the container image with deno only

The image previously shipped node_modules plus sources and ran tsx at
startup via a node base image. A single deno build stage installs the
runtime npm dependencies (devDependencies are stripped first so only
~10MB of packages get embedded) and compiles the binary; the final
stage is node-free and contains just git, git-lfs, wget, CA
certificates and the binary. scratch and alpine finals were evaluated
and rejected: deno compile has no musl target and the glibc binary
fails on alpine even with gcompat (__res_init missing), and the plugin
needs the git CLI, /bin/sh for hooks and CA certs at runtime.
Unverified with docker build (no docker daemon available); the build
stage was replayed manually from the exact copied file set and the
resulting binary smoke-tested.

claude committed:
feat: replace node toolchain with deno in the nix dev shell

deno covers runtime, package management, type checking and the
language server, so nodejs, pnpm, typescript and
typescript-language-server can go.

6543 commented on Dockerfile:
```diff
@@ -1,25 +1,31 @@
-RUN apt update \
-  && apt install -y git git-lfs wget \
+RUN apt-get update \
+  && apt-get install -y --no-install-recommends git git-lfs wget ca-certificates \
```
```suggestion
RUN apt update \
  && apt install -y --no-install-recommends git git-lfs wget ca-certificates \
```

6543 committed:
Apply suggestion from @6543
